### ISSUE TITLE
💬 RFC: History and cache in Backstage (also how to limit external API calls)

### ISSUE THREAD

**Issue by radoslaw-sz** [backstage/backstage#20472]
### 🔖 Need

Hi all,
I am working on integration of Backstage into some big organization and we have noticed two things, which can be problematic:
1) Intensive API calls to external systems
Organization contains many users, teams and components. We are using intensively three external systems - Jenkins, Gerrit and Spinnaker. At this moment all requests are being done when user opens a particular tab (plugin), so you may assume that there is at least couple of API calls for every system for every user who visits a page. We are worry about hitting API limits.
2) History
As mentioned, we are using three systems (Jenkins, Gerrit, Spinnaker) and we are merging information from them for the user. Unfortunately, Jenkins and Spinnaker has quite short history, so it happens often that we do not have proper result for some old Gerrit review. We are starting to think how to keep a history to be able to show users information even one year old. Plus we want to be sure that we have all information needed for Frontend plugin which uses 3 external sources of information.

I have noticed that there is no general solution for above problems. I checked Discord, Github issues and here are my findings:
1) Software Catalog has/had the first problem from above list - finally webhooks have been introduced https://github.com/backstage/backstage/issues/11082, which limits API calls. However, in my case, it is not about Software Catalog but about plugins.
2) Many people complain about reaching API limits and there was not solution for that. The only interesting one which I found is to create separated, external application which could cache history (see Alternatives).
3) No plugin uses Backstage Database to story history (or cache some info). Usage of Backstage Database is only for internal usage (like Announcement or Bazaar), not for scenario where Backstage gets info from external system and stores some of this info in DB.

### 🎉 Proposal

Proposal is just an idea for solving mentioned two problems, but it might break some architecture assumptions regarding Backstage. The whole RFC is rather to get a feedback to what direction we shall go, considering pros and cons.
Proposal contains solutions from https://github.com/backstage/backstage/issues/11082, so I consider Webhook, events and usage of database (like for Software Catalog). For sake of simplicity, let's consider Jenkins as a plugin which has above problems (but it could be any plugin or group of plugins which take info from external systems).
![image](https://github.com/backstage/backstage/assets/54472480/d33b43a2-c74b-4bc6-8e70-473f187279ed)

Here is a description of components **(component does not mean classes or modules - just a boundary of responsibilities):**
- _Jenkins_ represents Jenkins server/host
- _Jenkins webhook_ is responsible for exposing /POST endpoint for Jenkins notifications
- _Event backend_ is already existing module in Backstage
- _Jenkins DB handler_ responsible for subscribing for topics about Jenkins notifications and fills database. **Important note**: it does not do 1-1 mapping of API response to DB - it fills DB with only needed information for Frontend (or whatever you want)
- _Jenkins router_ - usual component which handles Frontend requests. However, instead of asking for requests to external system, it fetches info from DB (or external system - it can be configurable if needed).

Pros:
- Backstage relies on push model, instead of pull model also in case of external systems, so it is much easier to keep API calls on reasonable limit
- Backstage has cache and history in database which means that it can provide more stable information (for instance if external system does not respond, then Backstage plugin works. Anyway this situation provides much more functionalities).
- Solution is created for "push model", but it can be also used in "pull" one. It gives much more control, how Backstage uses external API - instead of _Jenkins webhook_ there can be component which limits API calls to external system or pulls info at some interval or some hybrid solution.
Cons:
- With more power, comes more responsibility. Backstage relies on internal database, so any mismatch needs to be handled/analyzed in Backstage - it can increase complexity of Backstage system.
- Push model can be used only if external system supports such way

### 〽️ Alternatives

I am not discarding this alternative, but I would like to get a feedback (or a decision) to which direction Backstage shall evolve. At this moment Backstage relies only on external systems - based on Proposal it still relies on them, but with much more control how data is being pulled into Backstage.
The alternative of solving mentioned problems is to create separated application which will do the same but outside of Backstage system. See picture below:
![image](https://github.com/backstage/backstage/assets/54472480/4360032e-8ec2-46aa-be99-0b18faca339d)

Here is a description of components **(component does not mean classes or modules - just a boundary of responsibilities):**
- _Jenkins_ represents Jenkins server/host
- _Jenkins webhook_ is responsible for exposing /POST endpoint for Jenkins notifications. It fills external database with information from Jenkins.
- _Jenkins custom router_ - it handles requests from Backstage. It fetches information from external DB.
- _Jenkins API_ - usual component which fetches info from external API.
- _Jenkins router_ usual component which handles Frontend requests.  It requests info from external API.

Pros:
- similar pros like in Proposal
- additionally implementation is done outside of Backstage (pros?)

Cons:
- custom application needs to be created for common issue in Backstage
- it can be open-sourced, but it won't be a part of Backstage ecosystem so it might be worse maintainable
- _Jenkins plugin_ needs to be configured to use this solution, unless there will be a way to completely abstract external API (I do not know if possible)
- people who struggle with intensive API calls problem (or history/cache problem) need to be directed to this solution

### ❌ Risks

Risks already mentioned in Proposal and Alternatives - open to gather more feedback on this.

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by taras** [comment]
Thank you for your proposal. I read through it, but I can't figure out what you're proposing. I recommend you describe the two options at a high level before getting into specifics. It could also help to give each option a name that briefly identifies the solution. The two approaches should be different from reading the title and description of each option.

**Comment by angeliski** [comment]
I like the concept, but I think the work in #13931 enables this at easy. In theory, you just need to create your EventSubscriber for the plugin (JenkinsEventSubscriber) and implement the logic to plugin check the database before make the request.

Just to check, why not talk about some cache system to avoid rate limit for some cases? For some plugin that don't have a webhook system in place for example

**Comment by radoslaw-sz** [comment]
Hi @taras 
Thanks the response. Let me maybe describe it in a brief.
**Problem:** As a big organization, we may have hundreds of users, who are using Backstage every second. Backstage calls external APIs very, very frequently (if user goes to plugin, calls happen) and we start to worry about reaching API limits. We would like to find a generic way, how to handle it. Additionally, external API provides limited number of information (e.g. last 50 actions), however we would like to show users information even 1 year old (so it could be 500th action from today in history).
**Solution inside Backstage**: This is a Proposal from the first message. We can use Webhooks, EventSubscriber and Backstage's database to limit API calls and keep the history.
**Solution outside Backstage**: This is Alternative from the first message. We can create external application which will use Webhooks and store information in its database. Backstage will use this external application as source of truth.

It is really high level description - details are mentioned in first message :)

**Comment by radoslaw-sz** [comment]
Hi @angeliski ,
Thanks for your comment - yeah, exactly it shall not be hard to implement it in Backstage, but I am wondering whether it is a good direction. I mean Backstage will start handle quite complex logic of keeping this information inside its database and so on. At this moment it is being done only for Software Catalog, so maybe it is something which can be considered also for other plugins.
And you are right regarding cache system - this is what I had in mind when writing "Alternative" in first message. Backstage can stay as it is, but we can create a system which will handle this caching + will control API calls. From Backstage pov the only difference is request to different API. But, as mentioned in "cons", we would need to do for every external API. Again - it is doable, but I am wondering if it is better/worse solution.

**Comment by angeliski** [comment]
@radoslaw-sz I am not talking about a different system to handle cache, but an integrated cache layer, something like cache aside or Read through ([ref](https://blog.bytebytego.com/p/top-caching-strategies)). We have this strategy for the processor:  https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/processing/DefaultCatalogProcessingOrchestrator.ts#L119.

So maybe we can provide a simple proxy wrap to init your plugin and use a cache.
Something like that:
```
jenkinsInfoProvider: new CacheStrategyApi(DefaultJenkinsInfoProvider.fromConfig({
    config: env.config,
    catalog,
  })),
```
And the complexity stay at the [CacheManager](https://github.com/backstage/backstage/blob/master/packages/backend-common/src/cache/CacheManager.ts).

Needs a deep down, but I think is a nice approach for the problem

**Comment by Silthus** [comment]
I think there could be more the the problem than caching. For example it is also not possible to add resilience, logging, tracing, etc. to outgoing requests made by other plugins or functions using `fetch`.

Maybe a service api around fetch could be added that would allow implementing a fetch service that supports middleware adding the above mentioned caching, but additionally provides options to add logging, tracing, circuit breakers and so forth. 

This would solve many problems with backstage making outgoing http requests to potentially uncontrollable systems.

**Comment by Rugvip** [comment]
> As a big organization, we may have hundreds of users, who are using Backstage every second. Backstage calls external APIs very, very frequently (if user goes to plugin, calls happen) and we start to worry about reaching API limits.

I want to challenge this assumption. I'd even say that it is often desirable to have direct user traffic (at least for this reason) to services rather than having backends make the requests. A couple of hundred users making requests every now and then should not be an issue for any upstream service. There is definitely a lot of complexity in the catalog backend where we're making a lot of requests on behalf of users to check for updates, and we've ended up building this system with caching and event based updates.

**Comment by radoslaw-sz** [comment]
Hi all,
Just to let you know - I think we are talking about two things in paralel, which is fine, but I would like to confirm.
1) We have request-based API, so when plugins fetches info based on user's activity
2) We have also (sometimes) option to use webhook and subscribe for a change from external API

2) is one of the options which can solve intensive usage of 1) (also gives some other possibilities) and this is my proposal, but my question is how to (and if) introduce Webhook-based API into Backstage (see pictures in first post) for plugins like for Software Catalog.

Intensive usage of request-based API can be solved in other ways (other than solution 2), which I think some of you mentioned.
And probably such discussion is also needed, as no all external APIs supports subscription mechanism.
But I would like to focus here on webhook (subscription) usage, and here is why (also it is an answer for @Rugvip message).

@Rugvip my scenario is quite simple - I would like to have near real-time information in Backstage and I do not want to flood external API with the same requests over and over again. In my organization there is no cloud - we are making API calls to our own servers which have limited capacity. And option to increase this capacity might be quite fast challenged. Infrastructure guys might ask - "Why you are making so many the same requests in one minute, while there is no change on the server?". The simple solution for that is subscription for a change.
My whole proposal is not about using the same code from catalog-backend - it is rather to use the same principle in plugins, so to store information in DB and subscribe for a change to heavily limit not needed external API calls. I can make it inside Backstage as part of separated plugin (**Proposal** in first post) or I can make separated application which will work as cache/database (outside of Backstage - **Alternative** in first post). For instance Jenkins or Gerrit or Spinnaker will have their own implemention how to cache/store in DB, because it can be quite specific. Finally maybe some kind of framework can be commonized.

I really appreciate your feedback - I am right now in a place where I am gathering input for analysis which I need to present to my client and choose the best solution. If I understand correctly you are rather saying that Software Catalog is an exception as solution itself might be very complex.

**Comment by radoslaw-sz** [comment]
@angeliski thanks for the proposal, but I am not sure how this cache could work... How can I know that nothing changed or something changed? I mean to decide whether I need to make new API call to external system or not?

**Comment by radoslaw-sz** [comment]
Hi All,
Any other thoughts regarding history/cache? Is making it in backstage can be an option or better to use some external history/caching system?

**Comment by Rugvip** [comment]
@radoslaw-sz we're at a point where we still don't understand the need for this tbh. I do see that it could be useful to place a cache in front of some services but I think there are plenty of existing options to use already, and are likely better deployed as part of that service rather than the Backstage deployment itself.

We do see the need for things in the direction of the history use-case but I don't think we're at the point where it would make sense to build a generic solution for it. We'd prefer to have a few real implementations of that in plugins first that we can refactor into something generic if that makes sense.

TL;DR: The concrete action moving forward for this I think should be to try this out in one or a few plugins to see if it makes sense

**Comment by radoslaw-sz** [comment]
@Rugvip  thank you very much for your feedback. That's fair, I need to still think about it, but if I will have concrete solution in the code for some plugin (PoC) then I will back to you.

Thanks all for the discussion, I am closing this ticket right now.
